align_system/utils/incontext_utils.py

```diff
@@ -0,0 +1,287 @@
+    '''
+    Abstract class for incontext example generator
+    Instances of this class have unique set_icl_datasets() functions for formatting prompt and reponses 
+    '''
+    @abstractmethod
+    def __init__(self,
+                 incontext_settings, 
+                 target_kdmas, 
+                 log,
+                 **kwargs):
+        self.incontext_settings = incontext_settings
+        self.target_kdmas = target_kdmas
+        self.log = log
+        self.set_icl_datasets()
+
+    @abstractmethod
+    def set_icl_datasets(self):
+        '''
+        Sets self.icl_datasets which contains all the ICL examples
+        This is specific to each instance of the class because the prompt and response will vary
+        read_icl_dataset_files() is a generic helper method for this step
+
+        The keys of self.icl_datasets are the 'kdma' keys from self.target_kdmas,
+        the values are a list of all ICL examples for that kdma,
+        each ICL example is a dictionary with keys: 'prompt', 'response'
+        For example: {kdma: [{'prompt':str, 'response':json}, ...], ...}
+        '''
+        self.icl_datasets = {}
+        pass
+
+    def read_icl_dataset_files(self):
+        '''
+        Helper function for set_icl_datasets() - reads dataset files and gets examples for target_kdmas
+        Returns incontext_data dictionary with format:
+            {kdma:[{state, actions, choices, kdma_values}, ...], ...}
+        '''
+        incontext_data = {}
+        # For each kdma
+        for target_kdma in self.target_kdmas:
+            sys_kdma_name = target_kdma['kdma']
+            # Add examples for each dataset file
+            dset_files = self.incontext_settings["datasets"][sys_kdma_name]
+            # If there is only one, make it a list for the following loop
+            if not isinstance(dset_files, list):
+                dset_files = [dset_files]
+            
+            # For each dataset file
+            for dset_f in dset_files:
+                with open(dset_f) as f:
+                    dset = json.load(f)
+                incontext_data[sys_kdma_name] = []
+                # Load each example in the dataset file
+                for icl_sample in dset:
+                    # Get state and actions
+                    state, actions = hydrate_scenario_state(icl_sample["input"])
+                    # Get choices
+                    choices = adm_utils.format_choices(
+                        [a.unstructured for a in actions],
+                        actions,
+                        state,
+                        self.log
+                    )
+                    # Get KDMA_values
+                    kdma_values = []
+                    for label in icl_sample["label"]:
+                        if sys_kdma_name not in label:
+                            kdma_values.append(None)
+                        else:
+                            kdma_values.append(label[sys_kdma_name])
```

Comment:
If we ever get to datasets with multiple KDMAs per action, this will strip that information away. Our whole ICL strategy is one-KDMA-at-a-time focused right now, so I think this is fine, but wanted to call it out in case we think it might be useful to keep that information around

Reply:
I don't think think this will strip away the info in cases where there are multiple KDMAs per action. We loop over target kdmas one at a time and add examples for each target KDMA to the dictionary separately. So if an example does not contain one KDMA value we set that value to None, but if it contains another KDMA value for another of the target KDMAs it will still retain that and add the example there. We should retain all ground truth KDMA values for the target KDMAs